<div class="container">
  <h4>Error handling strategies</h4>
  <div class="my-3" style="font-style: italic">Error generated randomly, open Network tab for monitoring</div>

  <div class="row mb-2">
    <button class="btn btn-primary" (click)="loadCourses1()">Error handling 0 - error callback</button>
    <span class="ml-2">returning Alternative Observer</span>
  </div>
  <div class="row mb-2">
      <button class="btn btn-primary" (click)="loadCourses1()">Error handling 1 - Alternative</button>
      <span class="ml-2">returning Alternative Observer</span>
  </div>
  <div class="row mb-2">
    <button class="btn btn-primary" (click)="loadCourses2()">Error handling 2 - Re-throw</button>
    <span class="ml-2">returning Error Observer</span>
  </div>
  <div class="row mb-2">
    <button class="btn btn-primary" (click)="loadCourses3()">Error handling 3 - Cleanup</button>
    <span class="ml-2">returning Error Observer, but closed with cleanup</span>
  </div>
  <div class="row mb-2">
    <button class="btn btn-primary" (click)="loadCourses4()">Error handling 4 - Retry</button>
    <span class="ml-2">on error retrying using Observable again</span>
  </div>

  <table class="table">
    <thead>
    <tr>
      <th>ID</th>
      <th>Description</th>
    </tr>
    </thead>
    <tbody>
    <tr *ngFor="let c of (courses$ | async)">
      <td>{{c.id}}</td>
      <td>{{c.description}}</td>
    </tr>
    </tbody>
  </table>
</div>
